hyperbolic-rs/rudders/losses.py

```diff
@@ -65,24 +69,78 @@ def loss_from_distances(self, dist_to_pos, dist_to_neg):
         return tf.reduce_mean(tf.nn.relu(self.margin - dist_to_pos + dist_to_neg))
 
 
-class HingeDistortionLoss(PairwiseHingeLoss):
+class SemanticLoss(LossFunction):
+    def __init__(self, n_users, n_items, args, **kwargs):
+        super(SemanticLoss, self).__init__(n_users, n_items, args)
+        self.item_distances = tf.keras.backend.constant(kwargs["item_distances"])
+        # Keep small proportion of closest neighbors of each item to compute loss only over those
+        # First set unreachable nodes to a large distance, then it looks for the indexes of min_k
+        # (top_k of -distances) to use them in the loss calculation
+        valid_item_dists = tf.where(self.item_distances > 0, self.item_distances, tf.keras.backend.ones(1) * 1000)
+        neighs_ids = tf.math.top_k(-valid_item_dists, k=int(len(self.item_distances) * args.neighbors))[1]
+        self.neighbor_ids = tf.cast(neighs_ids, tf.int64)
+        self.gamma = tf.Variable(args.semantic_gamma * tf.keras.backend.ones(1), trainable=False)
+        self.distortion_neg_sample_size = args.distortion_neg_sample_size
+
+    def calculate_loss(self, model, input_batch):
```

Comment:
Hopefully a tiny test for each loss would not be too hard implement?
People will trust your results more and it is often easier to look at the test to understand what the code is doing.

Reply:
Done